myia/type/unity.py

```diff
@@ -0,0 +1,173 @@
+        return value in self.legal_values
+
+
+class FilterVar(Var):
+    _props_names = ('tag', 'filter')
+    __slots__ = ('filter',)
+
+    def __init__(self, tag: str, filter: CallableT[[Type], bool]) -> None:
+        self.tag = tag
+        self.filter = filter
+
+    def matches(self, value: Type) -> bool:
+        if isinstance(value, RestrictedVar):
+            return all(self.filter(v) for v in value.legal_values)
+        if isinstance(value, FilterVar):
+            return self.filter == value.filter
+        return self.filter(value)
+
+
+FilterT = Union[Iterable[Type], CallableT[[Type], bool]]
+TagT = Union[str, FilterT]
+_SEQ = 1
+
+
+def var(tag: str = None, filter: FilterT = None) -> Var:
+    """
+    Create a variable for unification purposes.
+    Arguments:
+        filter: A predicate, or a set of values the variable is
+            allowed to take.
+    """
+    global _SEQ
+    if not isinstance(tag, str):
+        if filter is not None:
+            raise TypeError("tag must be a string")
+        filter = tag
+        tag = None
+    if tag is None:
+        tag = '_{}'.format(_SEQ)
+        _SEQ += 1
+    else:
+        if tag.startswith('_'):
+            raise ValueError('tag starting with _')
+    if callable(filter):
+        return FilterVar(tag, filter)
+    elif filter:
+        return RestrictedVar(tag, filter)
+    else:
+        return Var(tag)
+
+
+def unify_raw(w: Type, v: Type, equiv: DictT[Type, Type]) -> DictT[Type, Type]:
+    while w in equiv:
+        w = equiv[w]
+    while v in equiv:
+        v = equiv[v]
+
+    if isinstance(w, Var):
+        if w.matches(v):
+            equiv[w] = v
+            return equiv
+    if isinstance(v, Var):
+        if v.matches(w):
+            equiv[v] = w
+            return equiv
+
+    if w is v:
+        return equiv
+
+    if isinstance(w, List) and isinstance(v, List):
+        return unify_raw(w.element_type, v.element_type, equiv)
+
+    elif isinstance(w, Struct) and isinstance(v, Struct):
+        if sorted(w.elements.keys()) != sorted(v.elements.keys()):
+            raise TypeError("Mismatched Struct entries")
+        for k in w.elements.keys():
+            equiv = unify_raw(w.elements[k], v.elements[k], equiv)
+        return equiv
+
+    elif isinstance(w, Tuple) and isinstance(v, Tuple):
+        if len(w.elements) != len(v.elements):
+            raise TypeError("Tuples of differing length")
+        for we, ve in zip(w.elements, v.elements):
+            equiv = unify_raw(we, ve, equiv)
+        return equiv
+
+    elif isinstance(w, Callable) and isinstance(v, Callable):
+        if len(w.arguments) != len(v.arguments):
+            raise TypeError("Function with different number of arguments")
+        for we, ve in zip(w.arguments, v.arguments):
+            equiv = unify_raw(we, ve, equiv)
+        return unify_raw(w.retval, v.retval, equiv)
+
+    else:
+        raise TypeError("Type match error")
```

Comment:
I think a custom exception `UnificationError` would be safer to use (since you don't want to catch `TypeError`s that arise from genuine implementation bugs).